plugins/UM3NetworkPrinting/src/Cloud/CloudOutputDeviceManager.py

```diff
@@ -0,0 +1,142 @@
+# Cura is released under the terms of the LGPLv3 or higher.
+import json
+from threading import Timer
+from typing import Dict, Optional
+
+from PyQt5.QtNetwork import QNetworkRequest, QNetworkReply
+
+from UM.Logger import Logger
+from UM.Signal import Signal
+from cura.CuraApplication import CuraApplication
+from cura.NetworkClient import NetworkClient
+
+from .CloudOutputDevice import CloudOutputDevice
+from .Models import CloudCluster
+
+
+##  The cloud output device manager is responsible for using the Ultimaker Cloud APIs to manage remote clusters.
+#   Keeping all cloud related logic in this class instead of the UM3OutputDevicePlugin results in more readable code.
+#
+#   API spec is available on https://api.ultimaker.com/docs/connect/spec/.
+#
+class CloudOutputDeviceManager(NetworkClient):
+    
+    # The cloud URL to use for remote clusters.
+    API_ROOT_PATH = "https://api-staging.ultimaker.com/connect/v1"
+
+    # The interval with which the remote clusters are checked
+    CHECK_CLUSTER_INTERVAL = 5  # seconds
+    
+    def __init__(self):
+        super().__init__()
+
+        # Persistent dict containing the remote clusters for the authenticated user.
+        self._remote_clusters = {}  # type: Dict[str, CloudOutputDevice]
+
+        application = CuraApplication.getInstance()
+        self._output_device_manager = application.getOutputDeviceManager()
+        self._account = application.getCuraAPI().account
+        self._account.loginStateChanged.connect(self._getRemoteClusters)
+        
+        # When switching machines we check if we have to activate a remote cluster.
+        application.globalContainerStackChanged.connect(self._connectToActiveMachine)
+        
+        self._on_cluster_received = Signal()
+        self._on_cluster_received.connect(self._getRemoteClusters)
+
+
+    ##  Override _createEmptyRequest to add the needed authentication header for talking to the Ultimaker Cloud API.
+    def _createEmptyRequest(self, path: str, content_type: Optional[str] = "application/json") -> QNetworkRequest:
+        request = super()._createEmptyRequest(self.API_ROOT_PATH + path, content_type = content_type)
+        if self._account.isLoggedIn:
+            # TODO: add correct scopes to OAuth2 client to use remote connect API.
+            # TODO: don't create the client when not signed in?
+            request.setRawHeader(b"Authorization", "Bearer {}".format(self._account.accessToken).encode())
+        return request
+
+    ##  Gets all remote clusters from the API.
+    def _getRemoteClusters(self) -> None:
+        Logger.log("i", "Retrieving remote clusters")
+        if self._account.isLoggedIn:
+            self.get("/clusters", on_finished = self._onGetRemoteClustersFinished)
+
+        # Only start the polling thread after the user is authenticated
+        # The first call to _getRemoteClusters comes from self._account.loginStateChanged
+        timer = Timer(5.0, self._on_cluster_received.emit)
```

Comment:
Don't use Python timers. They are horribly unrealiable and also have the nice "habit" of messing up threads / event queue from qt. It's better to use QTimers.